packages/uploadthing/src/internal/handler.ts

```diff
@@ -223,135 +224,209 @@ export const buildRequestHandler = <
+                "x-uploadthing-version": UPLOADTHING_VERSION,
+              },
+            },
+          );
+
+          if (!uploadthingApiResponse.ok) {
+            console.error("[UT] unable to get presigned urls");
+            try {
+              const error = (await uploadthingApiResponse.json()) as unknown;
+              console.error(error);
+              return new UploadThingError({
+                code: "BAD_REQUEST",
+                cause: error,
+              });
+            } catch (cause) {
+              console.error("[UT] unable to parse response");
+              return new UploadThingError({
+                code: "URL_GENERATION_FAILED",
+                message: "Unable to get presigned urls",
+                cause,
+              });
+            }
+          }
+
+          // This is when we send the response back to the user's form so they can submit the files
+          const parsedResponse = (await uploadthingApiResponse.json()) as {
+            presignedUrl: { url: string; fields: Record<string, string> }; // ripped type from S3 package
+            name: string;
+            key: string;
+          }[];
+
+          if (process.env.NODE_ENV === "development") {
+            for (const file of parsedResponse) {
+              void conditionalDevServer(file.key);
+            }
+          }
+
+          return { body: parsedResponse, status: 200 };
+        }
+        case "failure": {
+          const { fileKey } = (await req.json()) as {
+            fileKey: string;
+          };
+
+          // Tell uploadthing to mark the upload as failed
+          const uploadthingApiResponse = await fetch(
+            generateUploadThingURL("/api/failureCallback"),
+            {
+              method: "POST",
+              body: JSON.stringify({
+                fileKey,
+              }),
+              headers: {
+                "Content-Type": "application/json",
+                "x-uploadthing-api-key": preferredOrEnvSecret,
+                "x-uploadthing-version": UPLOADTHING_VERSION,
+              },
+            },
+          );
+
+          if (!uploadthingApiResponse.ok) {
+            console.error("[UT] failed to mark upload as failed");
+            try {
+              const error = (await uploadthingApiResponse.json()) as unknown;
+              console.error(error);
+              return new UploadThingError({
+                message: "Failed to mark upload as failed",
+                code: "INTERNAL_SERVER_ERROR",
+                cause: error,
+              });
+            } catch (cause) {
+              console.error("[UT] unable to parse response");
+              return new UploadThingError({
+                code: "URL_GENERATION_FAILED",
+                message: "Unable to get presigned urls",
+                cause,
+              });
+            }
+          }
+
+          // Run the onUploadError callback
+          uploadable._def.onUploadError({
```

Comment:
Have we documented if this can throw or not? Might be good to tell people what this function should do. If this throws and isnt wrapped in try-catch errors might be weird?